crossplane/crossplane-runtime
# Add Vault as an External Secret Store

### Description of your changes

This PR introduces `Vault` as an External Secret Store implementing the interface introduced in #321.

- Supports both Vault KV Secret Engine v1 and v2.
- Initial support with Token Auth and Kubernetes Auth left as todo for a follow up PR

Fixes #316 

I have:


### How has this code been tested

Using [this](https://github.com/crossplane/provider-gcp/pull/417) manually crafted provider-gcp PR.

1. Create a kind cluster and deploy Vault using its helm chart.
2. Create a k8s secret with a Vault token.
3. Create the following StoreConfigs.

**For v1:**

```yaml
apiVersion: gcp.secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: vaultv1
spec:
  type: Vault
  defaultScope: crossplane-system
  vault:
    server: http://vault.vault-system:8200
    mountPath: kvv1/
    version: v1
    auth:
      method: Token
      token:
        source: Secret
        secretRef:
          namespace: crossplane-system
          name: vault-token
          key: token
```

**For v2:**

```yaml
apiVersion: gcp.secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: vaultv2
spec:
  type: Vault
  defaultScope: crossplane-system
  vault:
    server: http://vault.vault-system:8200
    mountPath: kvv2/
    version: v2
    auth:
      method: Token
      token:
        source: Secret
        secretRef:
          namespace: crossplane-system
          name: vault-token
          key: token
```

4. Create a GCP `ServiceAccount`.

5. Create the following `ServiceAccountKeys`:

```yaml
apiVersion: iam.gcp.crossplane.io/v1alpha1
kind: ServiceAccountKey
metadata:
  name: demo-serviceaccount-key
spec:
  forProvider:
    serviceAccountRef:
      name: perfect-test-sa
  publishConnectionDetailsTo:
    name: demo-serviceaccount-key
    metadata:
      labels:
        environment: development
        team: backend
      annotations:
        argocd.argoproj.io/sync-options: Prune=false
      type: Opaque
    configRef:
      #name: vaultv1
      name: vaultv2
```

6. Verify the ServiceAccountKey is synced and ready.
7. Verify connection secret created properly with all keys and metadata (if v2)

[contribution process]: https://git.io/fj2m9
